napari/_vispy/vendored/volume.py

```diff
@@ -959,6 +976,54 @@ def _in_loop_snippet(self):
     def _after_loop_snippet(self):
         return frag_dict[self.method]['after_loop']
 
+    @staticmethod
+    @lru_cache(maxsize=10)
+    def _build_clipping_planes_func(n_planes):
+        """Build the code snippet used to clip the volume based on self.clipping_planes."""
+        func = Function(
+            '$vars\nfloat clip_planes(vec3 loc, vec3 vol_shape) { float is_shown = 1.0; $clips; return is_shown; }')
+        # each plane is defined by a position and a normal vector
+        # the fragment is considered clipped if on the "negative" side of the plane
+        vars_template = '''
+            uniform vec3 u_clipping_plane_pos{idx};
+            uniform vec3 u_clipping_plane_norm{idx};
+            '''
+        clip_template = '''
+            vec3 relative_vec{idx} = loc - ( u_clipping_plane_pos{idx} / vol_shape );
+            float is_shown{idx} = dot(relative_vec{idx}, u_clipping_plane_norm{idx});
+            is_shown = min(is_shown{idx}, is_shown);
+            '''
+        all_vars = []
+        all_clips = []
+        for idx in range(n_planes):
+            all_vars.append(vars_template.format(idx=idx))
+            all_clips.append(clip_template.format(idx=idx))
+        func['vars'] = ''.join(all_vars)
+        func['clips'] = ''.join(all_clips)
+        return func
+
+    @property
+    def clipping_planes(self):
+        """Get the set of planes used to clip the volume.
+
+        Each plane is defined by a position and a normal vector (magnitude is irrelevant). Shape: (n_planes, 2, 3)
```

Comment:
should mention which side of the normal is kept and which is discarded, also should be clear about which index on axis 1 is for plane points and which for plane normals

Reply:
Uhm, this should probably be added upstream as well... Will do here on the napari side for now!